cli/internal/transformerpipeline/pipeline.go

```diff
@@ -106,22 +125,45 @@ type clientWrapper struct {
 	isClosed   bool
 }
 
-func (s clientWrapper) startBlocking() error {
+func (s *clientWrapper) startBlocking() error {
 	if s.nextSendFn == nil {
 		return errors.New("nextSendFn is nil")
 	}
-	for {
-		data, err := s.client.Recv()
-		if err == io.EOF {
-			err := s.nextClose()
-			return err
-		}
-		if err != nil {
-			return err
+
+	recvCh := make(chan *plugin.Transform_Request)
+	errCh := make(chan error)
+
+	// Recv can block forever (e.g. if transformer decides to), so
+	// we run it asynchronously and check if pipeline is closed every second.
+	go func() {
+		for {
+			data, err := s.client.Recv()
+			if err != nil {
+				errCh <- err
+			} else {
+				recvCh <- &plugin.Transform_Request{Record: data.Record}
+			}
 		}
-		if err := s.nextSendFn(
-			&plugin.Transform_Request{Record: data.Record},
-		); err != nil {
+	}()
+
+	for {
+		select {
+		case <-time.After(1 * time.Second): // Check if pipeline is closed every second
+			if s.isClosed {
+				return s.nextClose()
+			}
+		case req, ok := <-recvCh: // Propagate records to next transformer
+			if !ok {
```

Comment:
Nothing seems to close `recvCh` (so `ok` is never false?)

Reply:
That is a very good point. I removed this unreachable code.